How to freeze (#9984)
haskell/cabal: doc/how-to-freeze-versions.rst

performance issues, or some other kind of unexpected behaviour.  This is where
``cabal.project.freeze`` comes in, as it ensures that dependencies don't
unexpectedly change.  You can still update your dependencies, but you have to do
it on purpose, by modifying or by deleting and regenerating
``cabal.project.freeze`` file, and in the meantime you are guaranteed no
surprises will happen.

This consistency can be valuable as it ensures that all teammates, deployments,
and continuous integration are installing the exactly same dependencies.  So if
you are running and testing the code on your local machine, you are guaranteed
that your teammate and your continuous integration will be running the exact
same code, and that at the end that exact same code will get deployed.

A ``.freeze`` file can be good to have when developing for yourself or within a
private team.  If anyone using it can somehow have different inputs to
the solver then the ``.freeze`` file can be troublesome.  It can prevent the
solver from finding a different version of a dependency that would satisfy a
different architecture or a different compiler version and boot libraries.

.. Warning::

>     If publishing a package to Hackage, not matter what kind of component it

Reviewer: ```suggestion
    If publishing a package to Hackage, no matter what kind of component your project
```

Author: Thanks for spotting this typo @langston-barrett. I took the "no matter". The "it" here refers to the package and this, I kept.

    ensure that cabal will have newer versions to re-resolve the dependencies
    with. Alternatively, you can run ``cabal update
    --ignore-project``.

Freeze
    If the ``cabal.project.freeze`` file doesn't exist, generating one is a
    great way to see what versions of dependencies are currently being used even
    if you choose to discard the ``.freeze`` file after inspecting it.

Thaw, Freeze
    If you changed the version ranges of any of the dependencies in any of your
    project's package descriptions, in any ``.cabal`` file, then delete the
    ``cabal.project.freeze`` file if it already exists and run ``cabal freeze``
    to generate fresh version of ``cabal.project.freeze``.

Freeze, Freeze (Freezing Harder)
    You might in some cases want to skip deletion of ``cabal.project.freeze``,
    but keep in mind that in that case ``cabal freeze`` will use existing
    ``cabal.project.freeze`` when resolving dependencies, therefore not updating
    any existing dependencies, only adding new ones.

Partial Thaw, Freeze
    If you want to a pick up a different version of a single dependency, you can
    delete its constraint from ``cabal.project.freeze`` and then run ``cabal

>     freeze`` again.

Reviewer: I would suggest organizing these by the goals that the user is trying to solve, e.g.,

> Check which versions of dependencies Cabal has selected
>     If the ``cabal.project.freeze`` file doesn't exist, ...
>
> Update the versions of dependencies listed in the ``cabal.project.freeze`` file
>     To bump the versions of dependencies listed in the freeze file (for example, because you changed the version range in a ``.cabal`` file), first delete ...
>
> Add a new dependency to the ``cabal.project.freeze`` file
>     If you have added a new dependency to a ``.cabal`` file, simply re-run ...
>
> Updating a single dependency
>     ...

Author: Thanks @langston-barrett for this suggestion. I've redone this section on worflows.

Reviewer: Thanks, the new version is nice!

.. Note::

    If not sure, pick the "thaw, freeze" workflow, as it is the safest, the
    simplest and the most common. Finally, you will always want to commit the

>     changed ``cabal.project.freeze`` to version control.

Reviewer: > you will always want to commit the changed `cabal.project.freeze` to version control.

I find this a bit misleading - I believe you would want to commit the changed `cabal.project.freeze` iff one was already committed.

Author: Yes committing a freeze file is rarely done and I don't recommend it.

Ensuring everything is frozen
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. Note::

    If the ``.freeze`` file already has version equality constraints for every
    package that is a dependency of the project, then the solver will not be
    able to find a different version for any of them, the ``.freeze`` file
    cannot change and, at that point when every dependency is frozen, ``cabal
    freeze`` becomes an idempotent operation.

> Adding a dependency to one of the packages in a project without freezing harder

Reviewer: This line doesn't make sense unless you read the previous section to know what "freeze harder" means. Some readers might land on this section by following a link to it directly. Perhaps be more explicit?
```suggestion
Adding a dependency to one of the packages in a project without running ``cabal freeze`` again
```
or
```suggestion
Adding a dependency to one of the packages in a project without regenerating the ``cabal.project.freeze`` file using one of the methods described above
```
(Perhaps with a link to the above section)

Author: I've added a ref link to freezing harder in the previous section and also made bold the short descriptions; "cold snap", "thaw and freeze", "freeze harder" and "partial thaw and freeze".

leaves the newly added dependency susceptible to getting updated unexpectedly
when the solver can find a different version for it.  Running ``cabal freeze``
will show this vulnerability to a human or an automated check that notices a new
version equality constraint in the ``.freeze`` file, a constraint for a package

> that wasn't in the ``.freeze`` file before.

Reviewer: I would suggest rephrasing this paragraph along these lines

> When adding a new dependency to a project that utilizes a freeze file, it is often desirable to regenerate the freeze file using one of the methods described above. This ensures that the freeze file contains a constraint for the new dependency.

If we want to mention the idea of an automated check, perhaps explicitly suggest how one might work by providing shell commands that implement it.

Author: Thanks, I've used your suggested copy mostly as-is but made it a bit shorter.

> When adding a new dependency to a project that utilizes a freeze file
[#add-new-dep]_, regenerating the freeze file using one of the methods described
above ensures that the freeze file contains a constraint for the new dependency.

    ``index-state`` line from ``cabal.project`` (and from
    ``cabal.project.freeze`` if it exists) and then run ``cabal update`` to
    ensure that cabal will have newer versions to re-resolve the dependencies
    with. Alternatively, you can run ``cabal update
    --ignore-project``.

For the versions of all dependencies at once:

- To check solver-chosen versions - cold snap

    With this workflow we freeze, inspect then discard ``cabal.project.freeze``,
    always leaving the solver free to choose other versions.  This is great way
    to see what versions of dependencies are currently being used.

- To update versions of all dependencies - thaw and freeze

    If you changed the version ranges of any of the dependencies in any of your
    project's package descriptions, in any ``.cabal`` file, then delete the
    ``cabal.project.freeze`` file if it already exists and run ``cabal freeze``
    to generate fresh version of ``cabal.project.freeze``.  The steps of this
    workflow are delete and freeze. The solver is let free to choose versions of
    all dependencies but once it has decided, those versions are immediately
    pinned.

>     This is "thaw and freeze" workflow is the simplest way to work with a

Reviewer: ```suggestion
    This "thaw and freeze" workflow is the simplest way to work with a
```

    ``cabal.project.freeze`` file committed to source control.

For the version of a single dependency:
    
- To pin the version of a new dependency

    The steps of this workflow are add the new dependency and freeze. The solver

>     is free to chose its version for the new dependency while retaining the

Reviewer: ```suggestion
    is free to choose a version for the new dependency while retaining the
```

How to control versions
=======================

There are various ways and places to limit what versions the solver can pick for
dependencies.

Version ranges
    Within a package description, version ranges for dependencies can be; tight
    or loose or missing altogether.

Version constraints
    Within a project, version constraints for dependencies limit the versions
    that the solver can pick.

Mandatory version constraints
    With :cfg-field:`reject-unconstrained-dependencies` set to ``all``
    dependencies must have version constraints.

Curated version sets
    A project can import curated sets of packages and versions that are known to

>     work together, such as those Stackage provides for Cabal's use.

Reviewer 2: A link to a sample of `cabal.config` from Stackage would be useful for readers.

Capped repository versions
    Adding ``index-state`` to a project limits versions coming from Hackage to
    include only those that were available at the cutoff time.

Frozen versions
    Pins the versions picked by the solver for all dependencies. This is a way
    to preserve a set of versions found by the solver, a solver-curated set, if
    you will.

Version exceptions
    Allow newer or older dependencies.

Version ranges
--------------

A dependency can have a version range specified in the package description.
Published packages are expected to attach version ranges to dependencies that
conform to a set of rules that ``cabal check`` command will check for and report
on:

- that lower bounds are inclusive
- that upper bounds are exclusive
- that upper bounds don't have trailing zeros
- that upper bounds are not missing
- that upper bounds are not missing for ``base``

For large projects with many packages, it would be a lot of work to keep all
package dependency version ranges up to date. Ways of overcoming this are:

> - use a bash script to replace or update version ranges in all package descriptions

Reviewer 2: I mean, I've never seen such bash script. Does it really make sense even mentioning it as a solution?..

> - use a bash script to replace or update version ranges in all package descriptions

Author: If I'd not seen the bash solution the wild I wouldn't have mentioned it. I can't remember exactly how it was done but it was simple, a low-overhead "duct tape" solution that may have used comment delimiters around a section of `.cabal` files to be altered by the script.

I'm happy to not mention it or to reverse the order of the list.

I've just found out about `cabal-fmt` and `cabal-gild` [fragments](https://github.com/tfausak/cabal-gild#fragment) so will include a mention of those too.

        $ cabal freeze --with-compiler=ghc-8.10.7
        $ mv cabal.project.freeze ghc-8.10.7.freeze

        $ cabal freeze --with-compiler=ghc-9.10.1
        $ mv cabal.project.freeze ghc-9.10.1.freeze

        $ cabal freeze --with-compiler=ghc-9.12.1
        $ mv cabal.project.freeze ghc-9.12.1.freeze

        $ diff ghc-8.10.7.freeze ghc-9.12.1.freeze --unified
        -constraints: any.base ==4.14.3.0,
        -             any.ghc-prim ==0.6.1,
        -             any.integer-gmp ==1.0.3.0,
        -             any.rts ==1.0.1
        +constraints: any.base ==4.21.0.0,
        +             any.ghc-bignum ==1.3,
        +             any.ghc-internal ==9.1201.0,
        +             any.ghc-prim ==0.13.0,
        +             any.rts ==1.0.2

        $ diff ghc-9.10.1.freeze ghc-9.12.1.freeze --unified
        -constraints: any.base ==4.20.0.0,
        +constraints: any.base ==4.21.0.0,
                      any.ghc-bignum ==1.3,
        -             any.ghc-internal ==9.1001.0,
        -             any.ghc-prim ==0.11.0,
        +             any.ghc-internal ==9.1201.0,
        +             any.ghc-prim ==0.13.0,
                      any.rts ==1.0.2

> Do you need to freeze?

Author: Great suggestion, thanks. I've added a tip admonition for this to the "Project versions" section.

.. _how-to-control-versions:

How to control versions
=======================

There are various ways and places to limit what versions the solver can pick for
dependencies. Ways to control versions are:

:ref:`Version ranges <version-ranges>`

>     Within a package description, version ranges for dependencies can be; tight

Reviewer 2: Semicolon seems misplaced.

Author: Thanks, I removed the semicolon.

    Allow newer or older dependencies.

Places to control versions are the files and command line options related to
versioning but how to use these also depend on what we're building. If we're
building a product for its own sake then we're free to manage versions wherever
and however we please. If we're publishing a package to Hackage then that
platform places restrictions on how loose we can be when specifying
dependencies. Before publishing to Hackage, ``cabal check`` will warn about
loose versioning that Hackage will reject.

:ref:`Hackage versions <hackage-versions>`
    Not all dependency version ranges are accepted by Hackage. After publishing,
    a package on Hackage can be revised to change version ranges of its
    dependencies.

:ref:`Project versions <project-versions>`
    Almost any way and any place imaginable for setting versions is available
    when configuring a project.

.. _version-ranges:

Version ranges
--------------

A dependency can have a version range specified in the package description.
Published packages are expected to attach version ranges to dependencies that
conform to a set of rules that ``cabal check`` command will check for and report
on:

> - that lower bounds are inclusive

Reviewer 2: I'd suggest giving short examples here and below, like ">= 1.0 not > 1.0" and in a similar fashion.

- that upper bounds are exclusive
- that upper bounds don't have trailing zeros
- that upper bounds are not missing
- that upper bounds are not missing for ``base``

.. Note::

>     Missing an upper bound on ``base`` is more serious than it is for other

Reviewer 2: Perhaps "is treated by `cabal check` as being more serious".

- use a curated set of packages and versions and import these into the project

.. _version-constraints:

Version constraints
-------------------

Version constraints can be applied to the project or command line.

.. Note::

    While they do not directly restrain dependency versions, flag constraints
    can alter the set of dependencies of a package. This can happen when a
    dependency is behind a flag. For example, the `directory
    <https://hackage.haskell.org/package/directory-1.3.10.1/directory.cabal>`_
    package description has a flag that brings in the `os-string
    <https://hackage.haskell.org/package/os-string>`_ as a dependency and
    changes the version range for `filepath
    <https://hackage.haskell.org/package/filepath>`_:

    .. code-block:: cabal

        name: directory

        flag os-string
          description: Use the new os-string package
          default: False
          manual: False

        library
          if flag(os-string)
            build-depends: filepath >= 1.5.0.0, os-string >= 2.0.0
          else
            build-depends: filepath >= 1.4.100.0 && < 1.5.0.0

.. Warning::

>     Version constraints are additive. If I add a constraint it doesn't remove or

Reviewer 2: Saying "additive" is ambiguous: one could imagine that `build-depends: base < 4` followed by `build-depends: base > 5` is equivalent to `base < 4 || > 5`, adding the respective ranges, while actually it means `base < 4 && > 5`.

Author: I've removed "additive" and say a lot more about how the version constraints accumulate.

A good example here would be to supply version constraints via various means and then freeze to see the combined result.

    replace prior constraints on versions. Constraints don't have override
    semantics.

.. _mandated-versions:

Mandated version constraints
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Version constraints can be mandated with
:cfg-field:`reject-unconstrained-dependencies`.  When this flag is set to
``all``, each dependency must have some version constraint, even if this is just
``>0``. This flag can be made stricter and set to ``eq``, requiring each
dependency to have an equality version constraint.

.. _curated-versions:

Curated version sets
--------------------

Stackage provides curated sets of packages and versions that are known to work
together and are updated regularly.  The latest resolver is the nightly and this
typically lags a bit behind the latest available GHC version. The long term
support (LTS) resolvers will each follow a specific GHC version and are updated
less frequently.

- https://www.stackage.org/lts-24.34/cabal.config
- https://www.stackage.org/nightly-2026-03-25/cabal.config

These ``*.config`` files set the compiler version and warn about cabal not

> yet having revisions.

Reviewer 2: Cabal does have revisions?..

Author: We can't pin a dependency to a revision, #7833.

The Stack [unpack command](https://docs.haskellstack.org/en/stable/commands/unpack_command/#the-stack-unpack-command) shows the syntax it uses to pin a revision, `@rev:<number>` or `@sha256:<sha>`.

.. code-block::

    $ curl -s https://www.stackage.org/lts-24.34/cabal.config | grep "with-compiler"
    with-compiler: ghc-9.10.3

> If importing directly causes a version conflict, we recommend; downloading

Reviewer 2: Semicolon seems misplaced.

Author: Thanks for spotting that, I've removed it.

        -             any.ghc-prim ==0.11.0,
        +             any.ghc-internal ==9.1201.0,
        +             any.ghc-prim ==0.13.0,
                      any.rts ==1.0.2

Do you need to freeze?
^^^^^^^^^^^^^^^^^^^^^^

Why would you want to freeze? Don't we want to get minor updates of our
dependencies, or at least patches, as soon as we can?  Well, although they
shouldn't, it is possible that any kind of update introduces new bugs,
performance issues, or some other kind of unexpected behaviour.  This is where
``cabal.project.freeze`` comes in, as it ensures that dependencies don't
unexpectedly change.  You can still update your dependencies, but you have to do
it on purpose, by modifying or by deleting and regenerating
``cabal.project.freeze`` file, and in the meantime you are guaranteed no
surprises will happen.

This consistency can be valuable as it ensures that all teammates, deployments,
and continuous integration are installing the exactly same dependencies.  So if
you are running and testing the code on your local machine, you are guaranteed
that your teammate and your continuous integration will be running the exact

> same code, and that at the end that exact same code will get deployed.

Reviewer 2: That's assuming that all these environments run the same OS with the same set of system packages. 

BTW I don't know what happens if you `cabal freeze` on Linux so that `cabal.project.freeze` specifies a version of the `unix` package and try to build on Windows. Will it fall apart?
